Several `test_backup_restore_on_cluster` tests are broken

- test_backup_restore_on_cluster/test_cancel_backup.py::test_short_disconnection_doesnt_stop_restore
- test_backup_restore_on_cluster/test_cancel_backup.py::test_shutdown_cancels_backup
- test_backup_restore_on_cluster/test_cancel_backup.py::test_short_disconnection_doesnt_stop_backup

Link: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80115&sha=7e696bf2c6486d238ebfeddc330d6ae4e1eb211c&name_0=PR&name_1=Integration%20tests%20%28aarch64%2C%20distributed%20plan%2C%204%2F4%29

The 2nd common symptom for failure is that some pieces of a backup (not clear, either backup completed before CANCEL arrived or incomplete backup) are leftover and the 3 tests in `test_cancel_backup.py `fail with error message -
```
>           assert new_znodes == set()
E           AssertionError: assert {'backup-d82471b8-3b7e-478c-8340-52c4c814795f'} == set()
E             Extra items in the left set:
E             'backup-d82471b8-3b7e-478c-8340-52c4c814795f'
E             Full diff:
E             - set()
E             + {'backup-d82471b8-3b7e-478c-8340-52c4c814795f'}

```
This check and message is in the destructor of `class NoTrashChecker`.

When the test succeeds, we see messages like below. The key is the second-to-last message `BackupImpl: Removing all files of backup Disk('backups', 'aa08c832007143188196ca4cc22a6d6e')`

```
2025.07.24 12:09:19.279853 [ 20 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Debug> executeQuery: (from 172.16.1.1:59758) (query 1, line 1) BACKUP TABLE tbl ON CLUSTER 'cluster' TO Disk('backups', 'aa08c832007143188196ca4cc22a6d6e') SETTINGS id='aa08c832007143188196ca4cc22a6d6e' ASYNC (stage: Complete)
2025.07.24 12:09:19.279877 [ 20 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Test> CancellationChecker: Did not add the task because the timeout is 0, query_id: 177a7fa9-f13b-4f1a-9182-402f4427b0c7
2025.07.24 12:09:19.280031 [ 20 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Test> QueryMetricLog: Scheduling next collecting task for query_id 177a7fa9-f13b-4f1a-9182-402f4427b0c7 in 999 ms
2025.07.24 12:09:19.280363 [ 20 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Debug> executeQuery: Read 1 rows, 42.00 B in 0.000558 sec., 1792.1146953405018 rows/sec., 73.50 KiB/sec.
2025.07.24 12:09:19.280384 [ 20 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Debug> TCPHandler: Processed in 0.000845581 sec.
2025.07.24 12:09:19.376247 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupCoordinationOnCluster: Created start node #1 in ZooKeeper for the initiator (coordination version: 2)
2025.07.24 12:09:19.377297 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: alive|, num_hosts, started|
2025.07.24 12:09:19.381441 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupsWorker: Opening backup for writing
2025.07.24 12:09:19.381512 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupImpl: Writing backup: Disk('backups', 'aa08c832007143188196ca4cc22a6d6e')
2025.07.24 12:09:19.381640 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupsWorker: Opened backup for writing
2025.07.24 12:09:20.378223 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: The 'alive' node hasn't been updated in ZooKeeper for host node1:9000 for 1s
2025.07.24 12:09:20.386410 [ 681 ] {272e7e44-f756-46bc-9f1b-ac38b5537066} <Debug> executeQuery: (from 0.0.0.0:0, user: , initial_query_id: 177a7fa9-f13b-4f1a-9182-402f4427b0c7) /* ddl_entry=query-0000000001 */ BACKUP TABLE tbl TO Disk('backups', 'aa08c832007143188196ca4cc22a6d6e') SETTINGS id = 'aa08c832007143188196ca4cc22a6d6e', backup_uuid = 'b932fa7f-34c7-4bca-8430-403ebafd2c72', internal = true, async = true, host_id = 'node2:9000', cluster_host_ids = [['node1:9000', 'node2:9000']] (stage: Complete)
2025.07.24 12:09:20.390891 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: alive|node1:9000, started|node1:9000
2025.07.24 12:09:20.391331 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: The 'alive' node hasn't been updated in ZooKeeper for host node2:9000 for 1s
2025.07.24 12:09:20.393556 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: alive|node2:9000, started|node2:9000
2025.07.24 12:09:21.385166 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Waiting for host node1:9000 to finish
2025.07.24 12:09:21.398939 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: error
2025.07.24 12:09:21.399136 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> ProcessList: Cancelling the query (reason: CANCELLED_BY_USER)
2025.07.24 12:09:21.399394 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Error> BackupsWorker: Failed to make backup Disk('backups', 'aa08c832007143188196ca4cc22a6d6e'): Code: 394. DB::Exception: Got error from host node1:9000. DB::Exception: Query was cancelled. Stack trace:
2025.07.24 12:09:21.399414 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Warning> BackupImpl: An error happened, the backup won't be completed
2025.07.24 12:09:21.399425 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupCoordinationOnCluster: Sending exception from the initiator to other hosts: Code: 394. DB::Exception: Got error from host node1:9000. DB::Exception: Query was cancelled. (QUERY_WAS_CANCELLED)
2025.07.24 12:09:21.399435 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupCoordinationOnCluster: The error node already exists
2025.07.24 12:09:21.399440 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Waiting for host node1:9000 to finish after error before cleanup
2025.07.24 12:09:21.400138 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Waiting for host node1:9000 to finish after error before cleanup
2025.07.24 12:09:21.400367 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: finished|node1:9000
2025.07.24 12:09:21.400378 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected that some zookeeper nodes disappeared from the stage folder: alive|node1:9000
2025.07.24 12:09:21.400395 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Waiting for host node2:9000 to finish after error before cleanup
2025.07.24 12:09:21.402113 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Waiting for host node2:9000 to finish after error before cleanup
2025.07.24 12:09:21.402316 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: finished|node2:9000
2025.07.24 12:09:21.402326 [ 693 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Detected that some zookeeper nodes disappeared from the stage folder: alive|node2:9000
2025.07.24 12:09:21.402346 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Other hosts finished working on this backup
2025.07.24 12:09:21.402406 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Stopped the watching thread
2025.07.24 12:09:21.403505 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupCoordinationOnCluster: Created the 'finish' node in ZooKeeper for the initiator. Hosts [] haven't finished yet
2025.07.24 12:09:21.403526 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Information> BackupImpl: Removing all files of backup Disk('backups', 'aa08c832007143188196ca4cc22a6d6e')
2025.07.24 12:09:21.403809 [ 699 ] {177a7fa9-f13b-4f1a-9182-402f4427b0c7} <Trace> BackupCoordinationOnCluster: Removing nodes from ZooKeeper
```
The cancellation point for above is -
```
0. Poco::Exception::Exception(String const&, int) @ 0x000000001ec3c432
1. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000f813f9b
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000008e1602c
3. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x0000000008e202eb
4. DB::QueryStatus::throwQueryWasCancelled() const @ 0x000000001584c5e5
5. DB::QueryStatus::throwProperExceptionIfNeeded(unsigned long const&, unsigned long const&) @ 0x000000001584c515
6. DB::QueryStatus::checkTimeLimit() @ 0x000000001584cabd
7. DB::BackupEntriesCollector::setStage(String const&, String const&) @ 0x000000001430c14e
8. DB::BackupEntriesCollector::gatherMetadataAndCheckConsistency() @ 0x0000000014309851
9. DB::BackupEntriesCollector::run() @ 0x0000000014308554
10. DB::BackupsWorker::doBackup(std::shared_ptr<DB::IBackup>, std::shared_ptr<DB::ASTBackupQuery> const&, String const&, DB::BackupSettings const&, std::shared_ptr<DB::IBackupCoordination>, std::shared_ptr<DB::Context>, std::shared_ptr<DB::Context const> const&, bool, std::shared_ptr<DB::Cluster> const&) @ 0x000000001433980e
11. DB::BackupsWorker::BackupStarter::doBackup() @ 0x0000000014346b85
12. void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::BackupsWorker::startMakingBackup(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x0000000014341ecc
13. std::__packaged_task_func<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'(), std::allocator<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'()>, void ()>::operator()() @ 0x00000000118d1604
14. std::packaged_task<void ()>::operator()() @ 0x00000000118d1783
15. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::worker() @ 0x000000000f95d355
16. ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::Threa@@@
```

When the test fails, the "Removing all files ..." message is not seen. Messages when the test fails -
```
2025.07.12 20:01:10.150072 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Debug> executeQuery: (from 0.0.0.0:0, user: , initial_query_id: 3b20c256-e002-4e65-8568-4943e72826ac) /* ddl_entry=query-0000000020 */ BACKUP TABLE tbl TO Disk('backups', '076346902ef441dd88df29ea9b3cec35') SETTINGS id = '076346902ef441dd88df29ea9b3cec35', backup_uuid = 'd82471b8-3b7e-478c-8340-52c4c814795f', internal = true, async = true, host_id = 'node2:9000', cluster_host_ids = [['node1:9000', 'node2:9000']] (stage: Complete)
2025.07.12 20:01:10.150100 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Test> CancellationChecker: Did not add the task because the timeout is 0, query_id: 93887813-310b-4c29-9093-4f2678115889
2025.07.12 20:01:10.150240 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Test> QueryMetricLog: Scheduling next collecting task for query_id 93887813-310b-4c29-9093-4f2678115889 in 999 ms
2025.07.12 20:01:10.150282 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Test> ParallelFormattingOutputFormat: Parallel formatting is being used
2025.07.12 20:01:10.150697 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Debug> executeQuery: Read 1 rows, 55.00 B in 0.000698 sec., 1432.6647564469913 rows/sec., 76.95 KiB/sec.
2025.07.12 20:01:10.150789 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Debug> DDLWorker: Executed query: /* ddl_entry=query-0000000020 */ BACKUP TABLE tbl TO Disk('backups', '076346902ef441dd88df29ea9b3cec35') SETTINGS id = '076346902ef441dd88df29ea9b3cec35', backup_uuid = 'd82471b8-3b7e-478c-8340-52c4c814795f', internal = true, async = true, host_id = 'node2:9000', cluster_host_ids = [['node1:9000', 'node2:9000']]
2025.07.12 20:01:10.150803 [ 682 ] {93887813-310b-4c29-9093-4f2678115889} <Debug> MemoryTracker: Query peak memory usage: 4.00 MiB.
2025.07.12 20:01:10.155556 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Information> BackupCoordinationOnCluster: Created start node #2 in ZooKeeper for host node2:9000 (coordination version: 2)
2025.07.12 20:01:10.157299 [ 695 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: alive|, alive|node2:9000, num_hosts, started|, started|node2:9000
2025.07.12 20:01:10.163400 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> BackupsWorker: Opening backup for writing
2025.07.12 20:01:10.163472 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Information> BackupImpl: Writing backup: Disk('backups', '076346902ef441dd88df29ea9b3cec35')
2025.07.12 20:01:10.163559 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Information> BackupsWorker: Opened backup for writing
2025.07.12 20:01:11.149889 [ 695 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> BackupCoordinationOnCluster: Detected new zookeeper nodes appeared in the stage folder: error
2025.07.12 20:01:11.150134 [ 695 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> ProcessList: Cancelling the query (reason: CANCELLED_BY_USER)
2025.07.12 20:01:11.163713 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> BackupEntriesCollector: Will use path in backup: "/shards/1/replicas/2"
2025.07.12 20:01:11.163732 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> BackupEntriesCollector: Setting stage: gathering metadata (0)
2025.07.12 20:01:11.163861 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Error> BackupsWorker: Failed to make internal backup Disk('backups', '076346902ef441dd88df29ea9b3cec35'): Code: 394. DB::Exception: Got error from the initiator. DB::Exception: Query was cancelled. Stack trace:
2025.07.12 20:01:11.163894 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Warning> BackupImpl: An error happened, the backup won't be completed
2025.07.12 20:01:11.163910 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Information> BackupCoordinationOnCluster: Sending exception from host node2:9000 to other hosts: Code: 394. DB::Exception: Got error from the initiator. DB::Exception: Query was cancelled. (QUERY_WAS_CANCELLED)
2025.07.12 20:01:11.163920 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Information> BackupCoordinationOnCluster: The error node already exists
2025.07.12 20:01:11.163991 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Trace> BackupCoordinationOnCluster: Stopped the watching thread
2025.07.12 20:01:11.165446 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Information> BackupCoordinationOnCluster: Created the 'finish' node in ZooKeeper for host node2:9000, 1 hosts left
```
Note this message is present "BackupImpl: An error happened, the backup won't be completed", but the "Removing all files" message is not present. The cancellation point is -

```
2025.07.12 20:01:11.163861 [ 706 ] {93887813-310b-4c29-9093-4f2678115889} <Error> BackupsWorker: Failed to make internal backup Disk('backups', '076346902ef441dd88df29ea9b3cec35'): Code: 394. DB::Exception: Got error from the initiator. DB::Exception: Query was cancelled. Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000f8fd588
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000a693c1c
2. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x000000000a69edf4
3. DB::QueryStatus::throwQueryWasCancelled() const @ 0x00000000137232ec
4. DB::QueryStatus::throwProperExceptionIfNeeded(unsigned long const&, unsigned long const&) @ 0x0000000013723204
5. DB::BackupCoordinationStageSync::checkIfOtherHostsFinish(String const&, std::optional<std::chrono::duration<long long, std::ratio<1l, 1000l>>>, bool, bool&, bool) const @ 0x00000000125e9ee0
6. DB::BackupCoordinationStageSync::waitOtherHostsFinishImpl(String const&, std::optional<std::chrono::duration<long long, std::ratio<1l, 1l>>>, bool) const @ 0x00000000125e94dc
7. DB::BackupCoordinationStageSync::waitOtherHostsFinish(bool) const @ 0x00000000125e8e74
8. DB::BackupsWorker::BackupStarter::doBackup() @ 0x0000000012581690
9. void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<DB::BackupsWorker::startMakingBackup(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000001257fcf8
10. std::__packaged_task_func<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'(), std::allocator<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunnerUnsafe<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'()>, void ()>::operator()() @ 0x0000000010fc8814
11. std::packaged_task<void ()>::operator()() @ 0x0000000010fc89b8
12. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::worker() @ 0x000000000fa19464
13. void std::__function::__policy_invoker<void ()>::__call_impl[abi:ne190107]<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*>(void (ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool::*&&)(), ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::ThreadFromThreadPool*&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000fa1f07c
```
i am guessing that we reach here, but then don't delete the files -

```
        bool should_remove_files_in_backup = backup && backup_is_corrupted && backups_worker.remove_backup_files_after_failure
            && backup_coordination && backup_coordination->isErrorSet() &&
            (backup_coordination->isBackupQuerySentToOtherHosts() ? backup_coordination->allHostsFinished() : backup_coordination->finished());
```

--- a/tests/integration/test_backup_restore_on_cluster/test_cancel_backup.py
+++ b/tests/integration/test_backup_restore_on_cluster/test_cancel_backup.py
@@ -679,6 +679,9 @@ def test_short_disconnection_doesnt_stop_backup():
         assert get_num_system_processes(initiator, backup_id=backup_id) >= 1
 
         # Dropping connection for less than `failure_after_host_disconnected_for_seconds`
+        # When using faster_zk_disconnect_detect.xml (session_timeout_ms=5000),
+        # the drop duration must be short enough to avoid ZK session expiry.
+        max_drop_seconds = 1 if use_faster_zk_disconnect_detect else 4
         with PartitionManager() as pm:
             random_sleep(4)
             node_to_drop_zk_connection = random_node()
@@ -686,7 +689,7 @@ def test_short_disconnection_doesnt_stop_backup():
                 f"Dropping connection between {get_node_name(node_to_drop_zk_connection)} and ZooKeeper at {format_current_time()}"
             )
             pm.drop_instance_zk_connections(node_to_drop_zk_connection)
-            random_sleep(4)
+            random_sleep(max_drop_seconds)
             print(
                 f"Restoring connection between {get_node_name(node_to_drop_zk_connection)} and ZooKeeper at {format_current_time()}"
             )
@@ -733,6 +736,9 @@ def test_short_disconnection_doesnt_stop_restore():
         assert get_num_system_processes(initiator, restore_id=restore_id) >= 1
 
         # Dropping connection for less than `failure_after_host_disconnected_for_seconds`
+        # When using faster_zk_disconnect_detect.xml (session_timeout_ms=5000),
+        # the drop duration must be short enough to avoid ZK session expiry.
+        max_drop_seconds = 1 if use_faster_zk_disconnect_detect else 3
         with PartitionManager() as pm:
             random_sleep(3)
             node_to_drop_zk_connection = random_node()
@@ -740,7 +746,7 @@ def test_short_disconnection_doesnt_stop_restore():
                 f"Dropping connection between {get_node_name(node_to_drop_zk_connection)} and ZooKeeper at {format_current_time()}"
             )
             pm.drop_instance_zk_connections(node_to_drop_zk_connection)
-            random_sleep(3)
+            random_sleep(max_drop_seconds)
             print(
                 f"Restoring connection between {get_node_name(node_to_drop_zk_connection)} and ZooKeeper at {format_current_time()}"
             )

